### PR TITLE
Add chunk indices

### PR DESCRIPTION
**Pull request by ivanyu**

### CODE REVIEW

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndex.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndex.java:1]: Copyrights?

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndexWithoutCompressionTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class ChunkIndexWithoutCompressionTest {
+    @Test
+    void incorrectConstructorParams() {
+        assertThatThrownBy(() -> new ChunkIndexForFilesWithoutCompression(-1, 100, 100),
+            "Original chunk size cannot be negative, -1 given")
```
[the remark this answers is not in the document]
Comment (ivanyu) [reply commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndexWithoutCompressionTest.java]: Great, thanks! Fixed

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        this.chunksAdded += 1;
+    }
+
+    protected void addChunk0(final int transformedChunkSize) {}
+
+    @Override
+    public final void finish(final int finalTransformedChunkSize) {
+        if (finished) {
+            throw new IllegalStateException("Cannot finish already finished index");
+        }
```
Comment (ivanyu) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:62]: This whole mechanics around finishing an index asks for a separate builder class, but probably a bit later

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndexBuilderCommonTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        assertThatThrownBy(() -> index.findChunkForOriginalOffset(-1))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Offset must be non-negative, -1 given");
+    }
+
+    @Test
+    void findForEmptyFile() {
+        final ChunkIndex index = emptyFileChunkIndexBuilder.finish(0);
+
+        assertThat(index.findChunkForOriginalOffset(0)).isEqualTo(null);
```
Comment (AnatolyPopov) [inline commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkIndexBuilderCommonTest.java:109]: Nit:
```suggestion
        assertThat(index.findChunkForOriginalOffset(0)).isNull();
```

And below also

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ChunkTest {
+    @Test
+    void identical() {
+        final var c1 = new Chunk(0, 20, 30, 100, 120);
+        final var c2 = new Chunk(0, 20, 30, 100, 120);
+        assertThat(c1.equals(c2)).isTrue();
+        assertThat(c2.equals(c1)).isTrue();
+        assertThat(c1.hashCode()).isEqualTo(c2.hashCode());
```
Comment (AnatolyPopov) [inline commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java:30]:
```suggestion
        assertThat(c1).hasSameHashCodeAs(c2);
```
And there is also `doesNotHaveSameHashCodeAs` for other places

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    // it should not be persisted.
+    private final List<Chunk> chunks;
+
+    public VariableSizeChunkIndex(final int originalChunkSize,
+                                  final int originalFileSize,
+                                  final List<Integer> transformedChunks) {
+        super(originalChunkSize, originalFileSize, transformedChunks.get(transformedChunks.size() - 1));
+        this.transformedChunks = List.copyOf(Objects.requireNonNull(transformedChunks));
+
+        chunks = materializeChunks();
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:51]: From Effective Java:
> There are a few more restrictions that a class must obey to allow inheritance. Constructors must not invoke overridable methods, directly or indirectly. If you violate this rule, program failure will result. The superclass constructor runs before the subclass constructor, so the overriding method in the subclass will be invoked before the subclass constructor has run. If the overriding method depends on any initialization performed by the subclass constructor, the method will not behave as expected.

TLDR: we should not call overridable methods from constructors I think.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    // This only a materialization for convenience and performance,
+    // it should not be persisted.
+    private final List<Chunk> chunks;
+
+    public VariableSizeChunkIndex(final int originalChunkSize,
+                                  final int originalFileSize,
+                                  final List<Integer> transformedChunks) {
+        super(originalChunkSize, originalFileSize, transformedChunks.get(transformedChunks.size() - 1));
+        this.transformedChunks = List.copyOf(Objects.requireNonNull(transformedChunks));
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:49]: Something strange seems to be going on here, if we get to this line `transformedChunks` should not be null already I guess.

Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:49]: Also I think builder is not reusable, right? So it perhaps does not make sense to make a copy here since the constructor is called only in builder's `finish0`

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        chunks = materializeChunks();
+    }
+
+    public List<Chunk> chunks() {
+        return chunks;
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:55]: Instead of copying the list in the constructor I guess we would want to return a copy here to keep this list unmodifiable.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/FixedSizeChunkIndex.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    public FixedSizeChunkIndex(final int originalChunkSize,
+                               final int originalFileSize,
+                               final int transformedChunkSize,
+                               final int finalTransformedChunkSize) {
+        super(originalChunkSize, originalFileSize, finalTransformedChunkSize);
+
+        checkSize(transformedChunkSize, "Transformed chunk size");
+        this.transformedChunkSize = transformedChunkSize;
+
+        chunks = materializeChunks();
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/FixedSizeChunkIndex.java:53]: Same as for `VariableSizeChunkIndex` and overridable method calls in constructor.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/FixedSizeChunkIndex.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public List<Chunk> chunks() {
+        return chunks;
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/FixedSizeChunkIndex.java:57]: Same as in `VariableSizeChunkIndex` I would return a copy here.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 6 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:51]: I think this applies to the case where the `AbstractChunkIndex` constructor calls `materializeChunks` and its subsequent calls. However when `FixedSizeChunkIndex` and `VariableSizeChunkIndex` does this, it should be fine as all the necessary state is initialized. I can't find a better way to not repeat the `materializeChunks` code in both classes.

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ChunkTest {
+    @Test
+    void identical() {
+        final var c1 = new Chunk(0, 20, 30, 100, 120);
+        final var c2 = new Chunk(0, 20, 30, 100, 120);
+        assertThat(c1.equals(c2)).isTrue();
```
[the remark this answers is not in the document]
Comment (ivanyu) [reply commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java:28]: I kinda did this intentionally to show I'm checking the result of `equals` as a thing itself

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+public abstract class AbstractChunkIndexBuilder implements ChunkIndexBuilder {
+    protected final int originalChunkSize;
+    protected final int originalFileSize;
+
+    protected int chunksAdded = 0;
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java:23]: Does this need to be protected? It's only used in this class.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        if (remainOfOriginalFileSize() <= originalChunkSize) {
+            throw new IllegalStateException("This must be final chunk. Call `finish` instead.");
+        }
+
+        addChunk0(transformedChunkSize);
+
+        this.chunksAdded += 1;
+    }
+
+    protected void addChunk0(final int transformedChunkSize) {
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java:52]: Should this one be abstract?

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java (hunk 12 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java:23]: Nah, same for `finished`. Will make `private`.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:55]: I mean, `materializeChunks` produce already an unmodifiable list

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 7 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:49]: That's true, the builder is one-time. however I kept the check for basic hygiene. At some point soon we'll be deserializing those indices and also directly instantiating them in tests.

Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:49]: That's true, the builder is one-time. However, the same reason as for keeping `Objects.requireNonNull`

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+public abstract class AbstractChunkIndexBuilder implements ChunkIndexBuilder {
+    protected final int originalChunkSize;
+    protected final int originalFileSize;
+
+    protected int chunksAdded = 0;
+    protected boolean finished = false;
```
[the remark this answers is not in the document]
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndexBuilder.java:24]: I like `finished` more because "closing" doesn't seem to make much sense on a builder

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (AnatolyPopov) [reply commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java:28]: The difference is in the output of test failure. `isEqualTo` prints the actual objects side by side so it's easier to spot why the equality check is failing. Regular `isTrue` however will print only that test expected true but the result was false and imo this brings a little less value. But anyways this is nitpicking and you can leave it as it is if you want.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 6 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:51]: Maybe make `materializeChunks` final then? I understand your point and do agree but at the same time I can imagine that somebody might want to patch `FixedSizeChunkIndex` or `VariableSizeChunkIndex` for their use case using inheritance since the classes are not final also. So this is not a problem by itself right now but a potential problem in the future that will be hard to spot if there will be the case.

File: commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ivanyu) [reply commons/src/test/java/io/aiven/kafka/tiered/storage/commons/chunkindex/ChunkTest.java]: Ah nvm, did this change anyway

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 7 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:49]: As you wish but I do not see any value neither in copying it here nor making it unmodifiable since it's a private field.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:55]: Oh, sorry I was thinking about different field when mentioned constructor.
Regarding `materializeChunks` looks like I'm missing some context here. Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling `materializeChunks` from constructor and not from here directly? And moreover this *Index classes supposed to be immutable, correct?(This I guess should be mentioned somewhere for clarity)

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 7 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java]: Well, you're right. I removed the copying

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:57]: > Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling `materializeChunks` from constructor and not from here directly?

Yeah, the idea is to pre-calculate the chunks list and cache it. It makes sense because the chunk list is inevitable to be used and potentially multiple times. It's done in the constructor, because it's automatically thread-safe this way (no need to have locks/`volatile`).

> And moreover this *Index classes supposed to be immutable, correct?(This I guess should be mentioned somewhere for clarity)

Yes, you're right. I added this clarification.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        checkSize(originalFileSize, "Original file size");
+        this.originalFileSize = originalFileSize;
+
+        checkSize(finalTransformedChunkSize, "Final transformed chunk size");
+        this.finalTransformedChunkSize = finalTransformedChunkSize;
+    }
+
+    // Override with care, this method is indirectly called from a constructor.
+    protected final List<Chunk> materializeChunks() {
+        final List<Chunk> chunks = new ArrayList<>(chunkCount());
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:42]: `materializeChunks` is now final but `chunkCount` is not and I guess there is exactly the same issue with a call from constructor here.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        int originalPosition = 0;
+        int transformedPosition = 0;
+        if (chunkCount() == 0) {
+            chunks.add(new Chunk(0, 0, 0, 0, 0));
+        } else {
+            for (int chunkI = 0; chunkI < chunkCount(); chunkI++) {
+                final int originalSize = originalChunkSize(chunkI);
+                final int transformedSize = transformedChunkSize(chunkI);
```
Comment (AnatolyPopov) [inline commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:50]: Same thing as above.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:57]: > >Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling materializeChunks from constructor and not from here directly?
>
> Yeah, the idea is to pre-calculate the chunks list and cache it. It makes sense because the chunk list is inevitable to be used and potentially multiple times. It's done in the constructor, because it's automatically thread-safe this way (no need to have locks/volatile).

I'm starting to hesitate that it's a right approach taking into account all potential initialization issues. Probably we could use a similar approach but move this `materializeChunks` out of this hierarchy to some kind of wrapper maybe..

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java (hunk 15 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:42]: And AFAIU an option probably could be to make `chunkCount` a final field initialized in constructor instead since all the info is available in constructor to calculate it.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java (hunk 16 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:54]: What exactly?

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java (hunk 15 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java]: That's a good point, I made `chunkCount` a final field

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (ivanyu) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:58]: Extracting it in any form will not work as the algorithm inextricably depends on the object's state: in one way or another we'll have to pass closures on the state. The only way to make it "right" is to make two copies of `transformedChunkSize` and `originalChunkSize`, one to be called in the constructor and the other in other places, which doesn't make sense.

The book's advice is generally right, but it in this particular case there is no issue. The present code execution is correct and we left the warnings for the posterity in case the classes are to be modified or extended. I don't see how realistically it can fail.

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java (hunk 16 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java:54]: Was thinking about `transformedChunkSize` being overridable, but nevermind

File: commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java (hunk 8 above)
Comment (AnatolyPopov) [reply commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java:58]: Revisited it once again and with additions of finals seems to be fine now.

**Review by AnatolyPopov — APPROVED** [review]
LGTM
